warehouse/query-core/src/main/java/datawave/query/planner/DefaultQueryPlanner.java

```diff
@@ -3229,8 +3259,254 @@ public static Date getEndDateForIndexLookup(Date endDate) {
+            } finally {
+                stopwatch.stop();
+            }
+        }
+        return serializedIvaratorDirs;
+    }
+
+    protected Set<String> getIndexedFields() {
+        if (indexedFields == null && indexedFieldsCallable != null) {
+            TraceStopwatch stopwatch = stageStopWatch.newStartedStopwatch(indexedFieldsCallable.stageName());
+            try {
+                while (indexedFields == null) {
+                    indexedFields = indexedFieldsFuture.get(concurrentTimeoutMillis, TimeUnit.MILLISECONDS);
+                }
+                return indexedFields;
+            } catch (ExecutionException | InterruptedException | TimeoutException e) {
+                log.error("Failed to fetch indexed fields");
+                log.error(e.getStackTrace());
+                throw new RuntimeException(e);
+            } finally {
+                stopwatch.stop();
+            }
+        }
+
+        if (indexedFields == null) {
+            return Collections.emptySet();
+        }
+
+        return indexedFields;
+    }
+
+    protected Set<String> getIndexOnlyFields() {
+        if (indexOnlyFields == null && indexOnlyFieldsCallable != null) {
+            TraceStopwatch stopwatch = stageStopWatch.newStartedStopwatch(indexOnlyFieldsCallable.stageName());
+            try {
+                while (indexOnlyFields == null) {
+                    indexOnlyFields = indexOnlyFieldsFuture.get(concurrentTimeoutMillis, TimeUnit.MILLISECONDS);
+                }
+                return indexOnlyFields;
+            } catch (ExecutionException | InterruptedException | TimeoutException e) {
+                log.error("Failed to fetch index only fields");
+                log.error(e.getStackTrace());
+                throw new RuntimeException(e);
+            } finally {
+                stopwatch.stop();
+            }
+        }
+
+        if (indexOnlyFields == null) {
+            return Collections.emptySet();
+        }
+
+        return indexOnlyFields;
+    }
+
+    protected Set<String> getNonEventFields() {
+        if (nonEventFields == null && nonEventFieldsCallable != null) {
+            TraceStopwatch stopwatch = stageStopWatch.newStartedStopwatch(nonEventFieldsCallable.stageName());
+            try {
+                while (nonEventFields == null) {
+                    nonEventFields = nonEventFieldsFuture.get(concurrentTimeoutMillis, TimeUnit.MILLISECONDS);
+                }
+                return nonEventFields;
+            } catch (ExecutionException | InterruptedException | TimeoutException e) {
+                log.error("Failed to fetch non-event fields");
+                log.error(e.getStackTrace());
+                throw new RuntimeException(e);
+            } finally {
+                stopwatch.stop();
+            }
+        }
+
+        if (nonEventFields == null) {
+            return Collections.emptySet();
+        }
+
+        return nonEventFields;
+    }
+
```

Comment:
Would it be preferable to condense the preceeding eight methods to call a single getCallable() method? Seems like a bit of redundant try/catch and emptySet returning.

Reply:
I could certainly write a callable per return type, so a FieldSetCallable could handle several of these use cases. The others unfortunately need to remain because of the different return types.

Reply:
it might also be possible to return the callable rather than the value, this way you could wrap the waiting and error handling on a single method that is rused, then all of these would either have a clean get or throw a RuntimeException

Reply:
Common code for getting field sets is extracted to a common method in the next commit